File: backend-nodejs/src/modules/portfolioreview/services/portfolioReviewService.ts
import { Types } from "mongoose";
import { AppError } from "../../../common/utils/AppError";
import { logger } from "../../../common/utils/logger";
import { AssessmentStatus, ScoreCategoryCode } from "../../riskappetite/enums/riskEnums";
import { CreateRecommendationDto, EligibleFundResponseDto, FlowTypeResponseDto, PortfolioEntryResponseDto, PortfolioRecommendationResponseDto, PortfolioReviewResponseDto, RfItemResponseDto } from "../dto/portfolioDto";
import { EntryAction, FLOW_TYPE_DISPLAY_NAMES, RecommendationFlowType, RecommendationStatus, ReviewStatus } from "../enums/portfolioEnums";
import { EligibleFund, IEligibleFund } from "../models/EligibleFund";
import { IPortfolioRecommendation, PortfolioRecommendation } from "../models/PortfolioRecommendation";
import { IPortfolioEntry, IPortfolioReview, PortfolioReview } from "../models/PortfolioReview";
import { Client } from "../../customer/models/Client";
import { RiskAssessment } from "../../riskappetite/models/RiskAssessment";
import { portfolioPdfService } from "./portfolioPdfService";

export class PortfolioReviewService {
    // Mappers                                                     

    private mapToFundResponse(ef: IEligibleFund): EligibleFundResponseDto {
        return {
            id: ef._id.toString(),
            fundName: ef.fundName,
            isin: ef.isin,
            fundSubCategory: ef.fundSubCategory,
            assetClass: ef.assetClass,
            instrumentType: ef.instrumentType,
            scoreCategory: ef.scoreCategory || null,
        };
    }

    private mapToEntryResponse(pe: IPortfolioEntry): PortfolioEntryResponseDto {
        return {
            id: pe._id.toString(),
            fundName: pe.fundName,
            isin: pe.isin,
            units: pe.units,
            purchaseNav: pe.purchaseNav,
            currentNav: pe.currentNav,
            investedAmount: pe.investedAmount,
            currentValue: pe.currentValue,
            absReturnPct: pe.absReturnPct,
            gain: pe.gain,
            cagrPct: pe.cagrPct,
            holdingDays: pe.holdingDays,
            action: pe.action,
        };
    }

    private mapToReviewResponse(pr: IPortfolioReview): PortfolioReviewResponseDto {
        return {
            id: pr._id.toString(),
            clientId: pr.client.toString(),
            status: pr.status,
            totalInvested: pr.totalInvested,
            totalCurrentValue: pr.totalCurrentValue,
            totalGain: pr.totalGain,
            gainPercentage: pr.gainPercentage,
            cagr: pr.cagr,
            note: pr.note || null,
            entries: pr.entries.map((e) => this.mapToEntryResponse(e)),
            createdAt: pr.createdAt,
        };
    }

    private mapToRecommendationResponse(pr: IPortfolioRecommendation):
        PortfolioRecommendationResponseDto {
        const funds: RfItemResponseDto[] = pr.funds.map((item, index) => {
            const fund = item.eligibleFund as IEligibleFund;
            return {
                id: item._id ? item._id.toString() : String(index + 1),
                eligibleFund: this.mapToFundResponse(fund),
                amount: item.amount,
                replacesEntryId: item.replacesEntryId ? item.replacesEntryId.toString() : null,
                displayOrder: item.displayOrder,
            };
        });

        return {
            id: pr._id.toString(),
            clientId: pr.client.toString(),
            portfolioReviewId: pr.portfolioReview ? pr.portfolioReview.toString() : null,
            flowType: pr.flowType,
            status: pr.status,
            investorCategory: pr.investorCategory || null,
            generatedDocumentUrl: pr.generatedDocumentUrl || null,
            funds,
            createdAt: pr.createdAt,
        };
    }

    // Master Fund Universe 
    getFlowTypes(): FlowTypeResponseDto[] {
        return Object.values(RecommendationFlowType).map((flowType) => ({
            code: flowType,
            displayName: FLOW_TYPE_DISPLAY_NAMES[flowType]
        }));
    }

    async getEligibleFunds(categoryCode?: string): Promise<EligibleFundResponseDto[]> {
        const filter: Record<string, unknown> = { isActive: true };

        if (categoryCode && Object.values(ScoreCategoryCode).includes(categoryCode as ScoreCategoryCode)) {
            filter.scoreCategory = categoryCode;
        }

        const funds = await EligibleFund.find(filter).sort({ fundName: 1 });
        return funds.map((f) => this.mapToFundResponse(f));
    }

    // Portfolio Reviews and Holdings
    async getReview(reviewId: string): Promise<PortfolioReviewResponseDto> {
        const review = await PortfolioReview.findById(reviewId);
        if (!review) {
            logger.warn({ reviewId }, "Get review failed: Portfolio review not found");
            throw new AppError(`Portfolio review not found with id: ${reviewId}`, 404);
        }
        return this.mapToReviewResponse(review);
    }

    async getLatestReview(clientId: string): Promise<PortfolioReviewResponseDto> {
        const review = await PortfolioReview.findOne({
            client: new Types.ObjectId(clientId)
        }).sort({ createdAt: -1 });

        if (!review) {
            logger.warn({ clientId }, 'Get latest review failed: No review found for client');
            throw new AppError(`No portfolio review found for client: ${clientId}`, 404);
        }

        return this.mapToReviewResponse(review);
    }

    async getReviewHistory(clientId: string): Promise<PortfolioReviewResponseDto[]> {
        const reviews = await PortfolioReview.find({
            client: new Types.ObjectId(clientId),
        }).sort({ createdAt: -1 });

        return reviews.map((r) => this.mapToReviewResponse(r));
    }

    /**
     * Creates a realistic eCAS portfolio review session for testing
     * Contains 4 holdings: 2 marked HOLD, 2 marked SELL
     */
    async createSampleReview(clientId: string): Promise<PortfolioReviewResponseDto> {
        const client = await Client.findById(clientId);
        if (!client) {
            logger.warn({ clientId }, 'Create sample review failed: Client not found');
            throw new AppError('Client not found', 404);
        }

        const sampleEntries: IPortfolioEntry[] = [
            {
                _id: new Types.ObjectId(),
                fundName: 'HDFC Top 100 Fund',
                isin: 'INF179K01BE2',
                units: 1000.0,
                purchaseNav: 500.0,
                currentNav: 650.0,
                investedAmount: 500000.0,
                currentValue: 650000.0,
                gain: 150000.0,
                absReturnPct: 30.0,
                cagrPct: 14.5,
                holdingDays: 730,
                action: EntryAction.HOLD,
            },
            {
                _id: new Types.ObjectId(),
                fundName: 'Axis Bluechip Fund',
                isin: 'INF846K01164',
                units: 800.0,
                purchaseNav: 40.0,
                currentNav: 48.0,
                investedAmount: 32000.0,
                currentValue: 38400.0,
                gain: 6400.0,
                absReturnPct: 20.0,
                cagrPct: 12.0,
                holdingDays: 600,
                action: EntryAction.HOLD,
            },
            {
                _id: new Types.ObjectId(),
                fundName: 'Underperforming Infra Fund',
                isin: 'INF200K01INF',
                units: 500.0,
                purchaseNav: 100.0,
                currentNav: 90.0,
                investedAmount: 50000.0,
                currentValue: 45000.0,
                gain: -5000.0,
                absReturnPct: -10.0,
                cagrPct: -4.5,
                holdingDays: 800,
                action: EntryAction.SELL,
            },
            {
                _id: new Types.ObjectId(),
                fundName: 'High Expense Small Cap Fund',
                isin: 'INF109K01SMC',
                units: 400.0,
                purchaseNav: 120.0,
                currentNav: 125.0,
                investedAmount: 48000.0,
                currentValue: 50000.0,
                gain: 2000.0,
                absReturnPct: 4.16,
                cagrPct: 1.8,
                holdingDays: 850,
                action: EntryAction.SELL,
            },
        ];

        const review = await PortfolioReview.create({
            client: new Types.ObjectId(clientId),
            status: ReviewStatus.COMPLETED,
            totalInvested: 630000.0,
            totalCurrentValue: 783400.0,
            totalGain: 153400.0,
            gainPercentage: 24.35,
            cagr: 12.5,
            note: 'Sample eCAS upload parsed successfully',
            entries: sampleEntries,
        });

        logger.info({ reviewId: review._id, clientId }, 'Sample portfolio review session created');
        return this.mapToReviewResponse(review);
    }

    // Recommendations and Compliance Check
    async createRecommendation(request: CreateRecommendationDto): Promise<PortfolioRecommendationResponseDto> {
        // Mandatory compliance check - Client must have a completed risk assessment
        const latestAssessment = await RiskAssessment.findOne({
            client: new Types.ObjectId(request.clientId),
            status: AssessmentStatus.COMPLETED,
        }).sort({ completedAt: -1, createdAt: -1 });

        if (!latestAssessment) {
            logger.warn(
                { clientId: request.clientId },
                'Recommendation creation blocked: Client must complete a risk assessment first'
            );
            throw new AppError(
                `Client ${request.clientId} must complete a risk assessment before receiving recommendations`,
                400
            );
        }

        // Validate Strategy Flow
        let review: IPortfolioReview | null = null;
        if (request.flowType === RecommendationFlowType.REPLACE_FUNDS) {
            if (!request.portfolioReviewId) {
                logger.warn({ request }, 'Create recommendation failed: Missing portfolioReviewId for REPLACE_FUNDS');
                throw new AppError('portfolioReviewId is required for REPLACE_FUNDS flow', 400);
            }

            review = await PortfolioReview.findById(request.portfolioReviewId);
            if (!review) {
                logger.warn({ reviewId: request.portfolioReviewId }, 'Create recommendation failed: Review not found');
                throw new AppError(`Portfolio review not found: ${request.portfolioReviewId}`, 404);
            }

            if (review.client.toString() !== request.clientId) {
                logger.warn({ reviewClient: review.client, requestClient: request.clientId }, 'Review does not belong to client');
                throw new AppError(`Review does not belong to client: ${request.clientId}`, 400);
            }

            // Validate that every replaced entry exists and has action === SELL
            for (const item of request.funds) {
                if (item.replacesEntryId) {
                    const entry = review.entries.find((e) => e._id.toString() === item.replacesEntryId);
                    if (!entry) {
                        logger.warn({ entryId: item.replacesEntryId }, 'Replaced entry not found in review');
                        throw new AppError(`Holding ${item.replacesEntryId} does not belong to review ${review._id}`, 404);
                    }
                    if (entry.action !== EntryAction.SELL) {
                        logger.warn({ entryId: item.replacesEntryId, action: entry.action }, 'Holding is not marked SELL');
                        throw new AppError(
                            `Holding ${item.replacesEntryId} is marked ${entry.action} - only SELL holdings can be replaced`,
                            400
                        );
                    }
                }
            }
        }
        else if (request.flowType === RecommendationFlowType.NEW_PORTFOLIO) {
            for (const item of request.funds) {
                if (item.replacesEntryId) {
                    logger.warn({ item }, 'replacesEntryId provided for NEW_PORTFOLIO');
                    throw new AppError('replacesEntryId cannot be provided for NEW_PORTFOLIO flow', 400);
                }
            }
        }

        // Verify all eligible funds exist
        const fundItems = [];
        let order = 1;
        for (const item of request.funds) {
            const fund = await EligibleFund.findById(item.eligibleFundId);
            if (!fund) {
                logger.warn({ eligibleFundId: item.eligibleFundId }, 'Eligible fund not found');
                throw new AppError(`Eligible fund not found: ${item.eligibleFundId}`, 404);
            }

            fundItems.push({
                eligibleFund: fund._id,
                amount: item.amount,
                replacesEntryId: item.replacesEntryId ? new Types.ObjectId(item.replacesEntryId) : null,
                displayOrder: item.displayOrder || order++,
            });
        }

        // Create and save recommendation
        const recommendation = await PortfolioRecommendation.create({
            client: new Types.ObjectId(request.clientId),
            portfolioReview: review ? review._id : null,
            flowType: request.flowType,
            status: RecommendationStatus.SAVED,
            investorCategory: latestAssessment.scoreCategory?.code || null,
            funds: fundItems,
        });

        const populated = await PortfolioRecommendation.findById(recommendation._id).populate('funds.eligibleFund');

        logger.info({ recommendationId: recommendation._id, clientId: request.clientId }, 'Recommendation created');
        return this.mapToRecommendationResponse(populated || recommendation);
    }

    async getRecommendation(id: string): Promise<PortfolioRecommendationResponseDto> {
        const recommendation = await PortfolioRecommendation.findById(id).populate('funds.eligibleFund');
        if (!recommendation) {
            logger.warn({ id }, 'Get recommendation failed: Not found');
            throw new AppError(`Portfolio recommendation not found with id: ${id}`, 404);
        }
        return this.mapToRecommendationResponse(recommendation);
    }

    async getRecommendationsByClient(clientId: string): Promise<PortfolioRecommendationResponseDto[]> {
        const recommendations = await PortfolioRecommendation.find({
            client: new Types.ObjectId(clientId),
        })
            .populate('funds.eligibleFund')
            .sort({ createdAt: -1 });

        return recommendations.map((rec) => this.mapToRecommendationResponse(rec));
    }

    /**
     * Triggers non-blocking background PDF generation and returns current state (HTTP 202).
     */
    async triggerPdfGeneration(recommendationId: string): Promise<PortfolioRecommendationResponseDto> {
        const recommendation = await PortfolioRecommendation.findById(recommendationId).populate(
            'funds.eligibleFund'
        );

        if (!recommendation) {
            logger.warn({ recommendationId }, 'Trigger PDF generation failed: Not found');
            throw new AppError(`Portfolio recommendation not found: ${recommendationId}`, 404);
        }

        if (
            recommendation.status === RecommendationStatus.PDF_GENERATED &&
            recommendation.generatedDocumentUrl
        ) {
            return this.mapToRecommendationResponse(recommendation);
        }

        // Fire and forget non-blocking background PDF generation
        this.generatePdfAsync(recommendationId).catch((err) => {
            logger.error({ err, recommendationId }, 'Background PDF generation encountered an error');
        });

        return this.mapToRecommendationResponse(recommendation);
    }

    /**
     * Non-blocking background worker for PDF compilation.
     */
    async generatePdfAsync(recommendationId: string): Promise<void> {
        try {
            const rec = await PortfolioRecommendation.findById(recommendationId).populate('funds.eligibleFund');
            if (!rec) return;

            if (rec.status === RecommendationStatus.PDF_GENERATED && rec.generatedDocumentUrl) {
                return;
            }

            const documentUrl = await portfolioPdfService.generateRecommendationPdf(rec);

            rec.status = RecommendationStatus.PDF_GENERATED;
            rec.generatedDocumentUrl = documentUrl;
            await rec.save();

            logger.info({ recommendationId, documentUrl }, 'Background PDF compilation completed');
        } catch (ex) {
            logger.error({ ex, recommendationId }, 'Background PDF compilation failed');
            await PortfolioRecommendation.findByIdAndUpdate(recommendationId, {
                status: RecommendationStatus.PDF_FAILED,
            });
        }
    }
}

export const portfolioReviewService = new PortfolioReviewService();

File: backend-java/src/main/java/com/wealthtech/crm/modules/portfolioreview/service/PortfolioReviewService.java
package com.wealthtech.crm.modules.portfolioreview.service;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.wealthtech.crm.modules.portfolioreview.dto.CreateRecommendationRequest;
import com.wealthtech.crm.modules.portfolioreview.dto.EligibleFundResponse;
import com.wealthtech.crm.modules.portfolioreview.dto.FlowTypeResponse;
import com.wealthtech.crm.modules.portfolioreview.dto.GeneratedPdfResult;
import com.wealthtech.crm.modules.portfolioreview.dto.PortfolioEntryResponse;
import com.wealthtech.crm.modules.portfolioreview.dto.PortfolioRecommendationResponse;
import com.wealthtech.crm.modules.portfolioreview.dto.PortfolioReviewResponse;
import com.wealthtech.crm.modules.portfolioreview.dto.RfItemRequest;
import com.wealthtech.crm.modules.portfolioreview.dto.RfItemResponse;
import com.wealthtech.crm.modules.portfolioreview.entity.EligibleFund;
import com.wealthtech.crm.modules.portfolioreview.entity.PortfolioEntry;
import com.wealthtech.crm.modules.portfolioreview.entity.PortfolioRecommendation;
import com.wealthtech.crm.modules.portfolioreview.entity.PortfolioReview;
import com.wealthtech.crm.modules.portfolioreview.entity.RecommendationFundItem;
import com.wealthtech.crm.modules.portfolioreview.enums.EntryAction;
import com.wealthtech.crm.modules.portfolioreview.enums.RecommendationFlowType;
import com.wealthtech.crm.modules.portfolioreview.enums.RecommendationStatus;
import com.wealthtech.crm.modules.portfolioreview.enums.ReviewStatus;
import com.wealthtech.crm.modules.portfolioreview.repository.EligibleFundRepository;
import com.wealthtech.crm.modules.portfolioreview.repository.PortfolioEntryRepository;
import com.wealthtech.crm.modules.portfolioreview.repository.PortfolioRecommendationRepository;
import com.wealthtech.crm.modules.portfolioreview.repository.PortfolioReviewRepository;
import com.wealthtech.crm.modules.portfolioreview.repository.RecommendationFundItemRepository;
import com.wealthtech.crm.modules.riskappetite.entity.RiskAssessment;
import com.wealthtech.crm.modules.riskappetite.enums.AssessmentStatus;
import com.wealthtech.crm.modules.riskappetite.enums.ScoreCategory;
import com.wealthtech.crm.modules.riskappetite.repository.RaRepository;
import com.wealthtech.crm.infrastructure.s3.S3Service;
import com.wealthtech.crm.modules.usermanager.exception.NotFoundException;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

// GET EligibleFunds
// GET FlowTypes

// GET getPortfolioReview(Long reviewId)
// GET getLatestPortfolioReview(Long clientId)
// POST createSampleReview(Long clientId) - dummy endpoint

// POST createRecommendationRequest(CreateRecommendationRequest request)
// GET getRecommendationId(Long id)
// GET getRecommendationsByClient(Long clientId)
// POST generatePdf(Long id)

@Slf4j
@Service 
@RequiredArgsConstructor
public class PortfolioReviewService {
    private final PortfolioReviewRepository reviewRepo;
    private final EligibleFundRepository efRepo;
    private final PortfolioEntryRepository entryRepo;
    private final PortfolioRecommendationRepository recommendationRepo;
    private final RecommendationFundItemRepository fundItemRepository;
    private final RaRepository raRepo;
    private final PortfolioPdfGeneratorService pdfGeneratorService;
    private final S3Service s3Service;

    // ==========================================
    // 1. MASTER FUND UNIVERSE & UTILITIES
    // ==========================================

    public List<FlowTypeResponse> getFlowTypes() {
        return Arrays.stream(RecommendationFlowType.values())
                .map(type -> new FlowTypeResponse(type.name(), type.getDisplayName()))
                .toList();
    }

    public List<EligibleFundResponse> getEligibleFunds(String categoryCode) {
        List<EligibleFund> funds;
        if (categoryCode != null && !categoryCode.isBlank()) {
            ScoreCategory category = ScoreCategory.fromCode(categoryCode);
            funds = efRepo.findByScoreCategoryAndIsActiveTrue(category);
        } else {
            funds = efRepo.findAll().stream()
                    .filter(EligibleFund::getIsActive)
                    .toList();
        }
        return funds.stream().map(this::mapToFundResponse).toList();
    }

    // ==========================================
    // 2. PORTFOLIO REVIEWS & HOLDINGS (eCAS)
    // ==========================================

    public PortfolioReviewResponse getReview(Long reviewId) {
        PortfolioReview review = reviewRepo.findById(reviewId)
                .orElseThrow(() -> new NotFoundException("Portfolio review not found with id: " + reviewId));
        return mapToReviewResponse(review);
    }

    public PortfolioReviewResponse getLatestReview(Long clientId) {
        PortfolioReview review = reviewRepo.findTopByClientIdOrderByCreatedAtDesc(clientId)
                .orElseThrow(() -> new NotFoundException("No portfolio review found for client: " + clientId));
        return mapToReviewResponse(review);
    }

    public List<PortfolioReviewResponse> getReviewHistory(Long clientId) {
        return reviewRepo.findByClientIdOrderByCreatedAtDesc(clientId).stream()
                .map(this::mapToReviewResponse)
                .toList();
    }

    /**
     * Creates a realistic eCAS portfolio review session for testing.
     * Contains 4 holdings: 2 marked HOLD, 2 marked SELL.
     */
    @Transactional
    public PortfolioReviewResponse createSampleReview(Long clientId) {
        PortfolioReview review = PortfolioReview.builder()
                .clientId(clientId)
                .status(ReviewStatus.COMPLETED)
                .totalInvested(new BigDecimal("630000.00"))
                .totalCurrentValue(new BigDecimal("783400.00"))
                .totalGain(new BigDecimal("153400.00"))
                .gainPercentage(24.35)
                .cagr(12.5)
                .note("Sample eCAS upload parsed successfully")
                .build();

        PortfolioReview savedReview = reviewRepo.save(review);

        List<PortfolioEntry> entries = List.of(
                PortfolioEntry.builder()
                        .portfolioReview(savedReview)
                        .fundName("HDFC Top 100 Fund")
                        .isin("INF179K01BE2")
                        .units(new BigDecimal("1000.00"))
                        .purchaseNav(new BigDecimal("500.00"))
                        .currentNav(new BigDecimal("650.00"))
                        .investedAmount(new BigDecimal("500000.00"))
                        .currentValue(new BigDecimal("650000.00"))
                        .gain(new BigDecimal("150000.00"))
                        .absReturnPct(30.0)
                        .cagrPct(14.5)
                        .holdingDays(730)
                        .action(EntryAction.HOLD)
                        .build(),
                PortfolioEntry.builder()
                        .portfolioReview(savedReview)
                        .fundName("Axis Bluechip Fund")
                        .isin("INF846K01164")
                        .units(new BigDecimal("800.00"))
                        .purchaseNav(new BigDecimal("40.00"))
                        .currentNav(new BigDecimal("48.00"))
                        .investedAmount(new BigDecimal("32000.00"))
                        .currentValue(new BigDecimal("38400.00"))
                        .gain(new BigDecimal("6400.00"))
                        .absReturnPct(20.0)
                        .cagrPct(12.0)
                        .holdingDays(600)
                        .action(EntryAction.HOLD)
                        .build(),
                PortfolioEntry.builder()
                        .portfolioReview(savedReview)
                        .fundName("Underperforming Infra Fund")
                        .isin("INF200K01INF")
                        .units(new BigDecimal("500.00"))
                        .purchaseNav(new BigDecimal("100.00"))
                        .currentNav(new BigDecimal("90.00"))
                        .investedAmount(new BigDecimal("50000.00"))
                        .currentValue(new BigDecimal("45000.00"))
                        .gain(new BigDecimal("-5000.00"))
                        .absReturnPct(-10.0)
                        .cagrPct(-4.5)
                        .holdingDays(800)
                        .action(EntryAction.SELL)
                        .build(),
                PortfolioEntry.builder()
                        .portfolioReview(savedReview)
                        .fundName("High Expense Small Cap Fund")
                        .isin("INF109K01SMC")
                        .units(new BigDecimal("400.00"))
                        .purchaseNav(new BigDecimal("120.00"))
                        .currentNav(new BigDecimal("125.00"))
                        .investedAmount(new BigDecimal("48000.00"))
                        .currentValue(new BigDecimal("50000.00"))
                        .gain(new BigDecimal("2000.00"))
                        .absReturnPct(4.16)
                        .cagrPct(1.8)
                        .holdingDays(850)
                        .action(EntryAction.SELL)
                        .build()
        );

        entryRepo.saveAll(entries);
        savedReview.getEntries().addAll(entries);

        return mapToReviewResponse(savedReview);
    }

    // ==========================================
    // 3. RECOMMENDATIONS & COMPLIANCE GATE
    // ==========================================

    @Transactional
    public PortfolioRecommendationResponse createRecommendation(CreateRecommendationRequest request) {
        // MANDATORY COMPLIANCE GATE: Client must have completed a risk assessment
        RiskAssessment latestAssessment = raRepo
                .findTopByClientIdAndStatusOrderByCompletedAtDesc(request.clientId(), AssessmentStatus.COMPLETED)
                .orElseThrow(() -> new IllegalStateException(
                        "Client " + request.clientId() + " must complete a risk assessment before receiving recommendations"));

        RecommendationFlowType flowType;
        try {
            flowType = RecommendationFlowType.valueOf(request.flowType().trim().toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid flow_type: " + request.flowType());
        }

        PortfolioReview review = null;
        if (flowType == RecommendationFlowType.REPLACE_FUNDS) {
            if (request.portfolioReviewId() == null) {
                throw new IllegalArgumentException("portfolio_review_id is required for REPLACE_FUNDS flow");
            }
            review = reviewRepo.findById(request.portfolioReviewId())
                    .orElseThrow(() -> new NotFoundException("Portfolio review not found: " + request.portfolioReviewId()));

            if (!review.getClientId().equals(request.clientId())) {
                throw new IllegalArgumentException("Review does not belong to client: " + request.clientId());
            }

            // Validate that every replaced entry exists in this review and has action == SELL
            for (RfItemRequest item : request.funds()) {
                if (item.replacesEntryId() != null) {
                    PortfolioEntry entry = entryRepo.findById(item.replacesEntryId())
                            .orElseThrow(() -> new NotFoundException("Replaced entry not found: " + item.replacesEntryId()));
                    if (!entry.getPortfolioReview().getId().equals(review.getId())) {
                        throw new IllegalArgumentException("Holding " + item.replacesEntryId() + " does not belong to review " + review.getId());
                    }
                    if (entry.getAction() != EntryAction.SELL) {
                        throw new IllegalArgumentException("Holding " + item.replacesEntryId() + " is marked " + entry.getAction() + " - only SELL holdings can be replaced");
                    }
                }
            }
        } else if (flowType == RecommendationFlowType.NEW_PORTFOLIO) {
            for (RfItemRequest item : request.funds()) {
                if (item.replacesEntryId() != null) {
                    throw new IllegalArgumentException("replaces_entry_id cannot be provided for NEW_PORTFOLIO flow");
                }
            }
        }

        // Build and save parent Recommendation
        PortfolioRecommendation recommendation = PortfolioRecommendation.builder()
                .clientId(request.clientId())
                .portfolioReview(review)
                .flowType(flowType)
                .status(RecommendationStatus.SAVED)
                .investorCategory(latestAssessment.getScoreCategory())
                .build();

        PortfolioRecommendation savedRecommendation = recommendationRepo.save(recommendation);

        // Build and save child fund line-items
        List<RecommendationFundItem> items = new ArrayList<>();
        int order = 1;
        for (RfItemRequest itemReq : request.funds()) {
            EligibleFund fund = efRepo.findById(itemReq.eligibleFundId())
                    .orElseThrow(() -> new NotFoundException("Eligible fund not found: " + itemReq.eligibleFundId()));

            PortfolioEntry replacesEntry = null;
            if (itemReq.replacesEntryId() != null) {
                replacesEntry = entryRepo.findById(itemReq.replacesEntryId()).orElse(null);
            }

            RecommendationFundItem fundItem = RecommendationFundItem.builder()
                    .recommendation(savedRecommendation)
                    .eligibleFund(fund)
                    .amount(itemReq.amount())
                    .replacesEntry(replacesEntry)
                    .displayOrder(itemReq.displayOrder() != null ? itemReq.displayOrder() : order++)
                    .build();

            items.add(fundItem);
        }

        fundItemRepository.saveAll(items);
        savedRecommendation.getFunds().addAll(items);

        return mapToRecommendationResponse(savedRecommendation);
    }

    public PortfolioRecommendationResponse getRecommendation(Long id) {
        PortfolioRecommendation recommendation = recommendationRepo.findById(id)
                .orElseThrow(() -> new NotFoundException("Portfolio recommendation not found with id: " + id));
        return mapToRecommendationResponse(recommendation);
    }

    public List<PortfolioRecommendationResponse> getRecommendationsByClient(Long clientId) {
        return recommendationRepo.findByClientIdOrderByCreatedAtDesc(clientId).stream()
                .map(this::mapToRecommendationResponse)
                .toList();
    }

    /**
     * Non-blocking asynchronous PDF generation running on Spring's @Async thread pool.
     */
    @Async
    public void generatePdfAsync(Long recommendationId) {
        try {
            PortfolioRecommendation rec = recommendationRepo.findById(recommendationId)
                    .orElseThrow(() -> new NotFoundException("Recommendation not found: " + recommendationId));

            if (rec.getStatus() == RecommendationStatus.PDF_GENERATED && rec.getGeneratedDocumentUrl() != null) {
                return;
            }

            GeneratedPdfResult result = pdfGeneratorService.generateRecommendationPdf(rec);

            rec.setStatus(RecommendationStatus.PDF_GENERATED);
            rec.setGeneratedDocumentUrl(result.presignedUrl());
            rec.setDocumentS3Key(result.s3Key());
            recommendationRepo.save(rec);

        } catch (Exception ex) {
            recommendationRepo.findById(recommendationId).ifPresent(rec -> {
                rec.setStatus(RecommendationStatus.PDF_FAILED);
                recommendationRepo.save(rec);
            });
        }
    }

    /**
     * Triggers async generation and returns the current recommendation state immediately (HTTP 202).
     */
    public PortfolioRecommendationResponse triggerPdfGeneration(Long recommendationId) {
        PortfolioRecommendation recommendation = recommendationRepo.findById(recommendationId)
                .orElseThrow(() -> new NotFoundException("Portfolio recommendation not found: " + recommendationId));

        if (recommendation.getStatus() == RecommendationStatus.PDF_GENERATED && recommendation.getGeneratedDocumentUrl() != null) {
            return mapToRecommendationResponse(recommendation);
        }

        generatePdfAsync(recommendationId);

        return mapToRecommendationResponse(recommendation);
    }

    // Mappers
    private EligibleFundResponse mapToFundResponse(EligibleFund ef) {
        return new EligibleFundResponse(
            ef.getId(), 
            ef.getFundName(), 
            ef.getIsin(), 
            ef.getFundSubCategory(), 
            ef.getAssetClass(), 
            ef.getInstrumentType(), 
            ef.getScoreCategory() != null ? ef.getScoreCategory().getCode() : null
        );
    }

    private PortfolioEntryResponse mapToEntryResponse(PortfolioEntry pe) {
        return new PortfolioEntryResponse(
            pe.getId(),
            pe.getFundName(),
            pe.getIsin(),
            pe.getUnits(),
            pe.getPurchaseNav(),
            pe.getCurrentNav(),
            pe.getInvestedAmount(),
            pe.getCurrentValue(),
            pe.getAbsReturnPct(),
            pe.getGain(),
            pe.getCagrPct(),
            pe.getHoldingDays(),
            pe.getAction() != null ? pe.getAction().name() : null
        );
    }

    private PortfolioReviewResponse mapToReviewResponse(PortfolioReview pr) {
        List<PortfolioEntryResponse> entries = pr.getEntries().stream()
                .map(this::mapToEntryResponse)
                .toList();

        return new PortfolioReviewResponse(
            pr.getId(),
            pr.getClientId(),
            pr.getStatus() != null ? pr.getStatus().name() : null,
            pr.getTotalInvested(),
            pr.getTotalCurrentValue(),
            pr.getTotalGain(),
            pr.getGainPercentage(),
            pr.getCagr(),
            pr.getNote(),
            entries,
            pr.getCreatedAt()
        );
    }

    private PortfolioRecommendationResponse mapToRecommendationResponse(PortfolioRecommendation pr) {
        String docUrl = pr.getGeneratedDocumentUrl();
        if (pr.getDocumentS3Key() != null && !pr.getDocumentS3Key().isBlank()) {
            try {
                docUrl = s3Service.generatePresignedGetUrl(pr.getDocumentS3Key());
            } catch (Exception e) {
                log.warn("Failed to generate presigned URL for documentS3Key {}: {}", pr.getDocumentS3Key(), e.getMessage());
            }
        }

        List<RfItemResponse> funds = pr.getFunds().stream()
                .map(item -> new RfItemResponse(
                    item.getId(),
                    mapToFundResponse(item.getEligibleFund()),
                    item.getAmount(),
                    item.getReplacesEntry() != null ? item.getReplacesEntry().getId() : null,
                    item.getDisplayOrder() 
                ))
                .toList();

        return new PortfolioRecommendationResponse(
            pr.getId(),
            pr.getClientId(),
            pr.getPortfolioReview() != null ? pr.getPortfolioReview().getId() : null,
            pr.getFlowType() != null ? pr.getFlowType().name() : null,
            pr.getStatus() != null ? pr.getStatus().name() : null,
            pr.getInvestorCategory() != null ? pr.getInvestorCategory().getCode() : null,
            docUrl,
            pr.getDocumentS3Key(),
            funds,
            pr.getCreatedAt()
        );
    }
}
